scipy/ndimage/_measurements.py

```diff
@@ -307,6 +307,115 @@ def find_objects(input, max_label=0):
     return _nd_image.find_objects(input, max_label)
 
 
+def value_indices(arr, *, ignore_value=None):
+    """
+    Find indices of each distinct value in given array.
+
+    Parameters
+    ----------
+    arr : ndarray of ints
+        Array containing integer values.
+    ignore_value : int, optional
+        This value will be ignored in searching the `arr` array. If not
+        given, all values found will be included in output. Default
+        is None.
+
+    Returns
+    -------
+    indices : dictionary
+        A Python dictionary of array indices for each distinct value. The
+        dictionary is keyed by the distinct values, the entries are array
+        index tuples covering all occurrences of the value within the
+        array.
+
+        This dictionary can occupy significant memory, usually several times
+        the size of the input array.
+
+    Notes
+    -----
+    For a small array with few distinct values, one might use
+    `numpy.unique()` to find all possible values, and ``(arr == val)`` to
+    locate each value within that array. However, for large arrays,
+    with many distinct values, this can become extremely inefficient,
+    as locating each value would require a new search through the entire
+    array. Using this function, there is essentially one search, with
+    the indices saved for all distinct values.
+
+    This is useful when matching a categorical image (e.g. a segmentation
```

Comment:
FWIW, here's roughly the equivalent of this function using scikit-image's regionprops:

```
from skimage import measure

def value_indices(a):
    return {r.label: r.coords for r in measure.regionprops(a)}
```

Not as efficient, of course.

Reply:
Wow, thank you for that. I was unaware of that `regionprops` function. Good to know about. You are right about its effficiency, though. I just did some testing, and while the initial call to `regionprops()` is very fast, the loop to get coordinates degrades very badly with increased number of distinct values, and also with image size. It also seems to have some surprising internal memory behaviour, my first test was "Killed" (running on Linux), for using too much memory, and I could only run fairly small examples to completion. 

I am very grateful that you mentioned it, though. I originally wrote my first version of this around 2009 (that one used scipy.weave), and have always felt guilty for not contributing it back. With a couple of intermediate versions in between (weave -> f2py, then a standalone extension module), I always thought I had not searched hard enough search to see if anyone had done something better in the meantime. Your mention of regionprops does suggest that I am not entirely re-inventing someone else's work. I still worry about this :-)

Reply:
P.S. On that "surprising memory behaviour" in regionprops. I just wandered through the source, and it turns out that the toy examples I gave it were "worst case" for that approach, with non-contiguous patches of each value. Would not be a problem in more normal use. :-)

Reply:
Oh, good, thanks for investigating! Yes, so in principle the operations should be reasonably effective. Each regionprop stores a link to the label image and a window of interest. When you query a regionprop, it extracts the relevant window of the label image, and then returns the coordinates of its label in that window. That's why, if you have a bunch of single pixel areas, the performance suffers badly. However, overall memory consumption should be quite low.